src/ncp/hdlc.cpp

```diff
@@ -103,76 +103,101 @@ uint16_t UpdateFcs(uint16_t aFcs, uint8_t aByte)
     return (aFcs >> 8) ^ sFcsTable[(aFcs ^ aByte) & 0xff];
 }
 
-ThreadError Encoder::Init(uint8_t *aOutBuf, uint16_t &aOutLength)
+
+Encoder::BufferWriteIterator::BufferWriteIterator(void)
+{
+    mWritePointer = NULL;
+    mRemainingLength = 0;
+}
+
+ThreadError Encoder::BufferWriteIterator::WriteByte(uint8_t aByte)
 {
     ThreadError error = kThreadError_None;
 
-    mFcs = kInitFcs;
+    VerifyOrExit(mRemainingLength > 0, error = kThreadError_NoBufs);
 
-    VerifyOrExit(aOutLength > 0, error = kThreadError_NoBufs);
-    aOutBuf[0] = kFlagSequence;
-    aOutLength = 1;
+    *mWritePointer++ = aByte;
+    mRemainingLength--;
 
 exit:
     return error;
 }
 
-ThreadError Encoder::Encode(uint8_t aInByte, uint8_t *aOutBuf, uint16_t aOutLength)
+bool Encoder::BufferWriteIterator::CanWrite(uint16_t aWriteLength)
 {
-    ThreadError error = kThreadError_None;
+   return (mRemainingLength >= aWriteLength);
+}
 
-    mFcs = UpdateFcs(mFcs, aInByte);
+
+ThreadError Encoder::Init(BufferWriteIterator &aIterator)
+{
+    mFcs = kInitFcs;
+
+    return aIterator.WriteByte(kFlagSequence);
+}
+
+ThreadError Encoder::Encode(uint8_t aInByte, BufferWriteIterator &aIterator)
+{
+    ThreadError error = kThreadError_None;
 
     if (aInByte == kFlagSequence || aInByte == kEscapeSequence)
     {
-        VerifyOrExit(mOutOffset + 2 < aOutLength, error = kThreadError_NoBufs);
-        aOutBuf[mOutOffset++] = kEscapeSequence;
-        aOutBuf[mOutOffset++] = aInByte ^ 0x20;
+        VerifyOrExit(aIterator.CanWrite(2) , error = kThreadError_NoBufs);
+
+        aIterator.WriteByte(kEscapeSequence);
+        aIterator.WriteByte(aInByte ^ 0x20);
     }
     else
     {
-        VerifyOrExit(mOutOffset + 1 < aOutLength, error = kThreadError_NoBufs);
-        aOutBuf[mOutOffset++] = aInByte;
+        SuccessOrExit(aIterator.WriteByte(aInByte));
```

Comment:
Set error here?  For example: `SuccessOrExit(error = aIterator.WriteByte(aInByte));`